File: bench/vec3.go
// Go twin of vec3.ts: the same methods on a heap-allocated *Vec3 (like the
// arena objects in Nish and the Box<Vec3> in the Rust twin), the same
// expression order. Nothing is allocated inside the loop.
package main

import (
	"fmt"
	"math"
)

const N = 50000000 // bench:n

type Vec3 struct {
	x, y, z float64
}

func vec3(x, y, z float64) *Vec3 {
	return &Vec3{x: x, y: y, z: z}
}

func (t *Vec3) add(o *Vec3) {
	t.x = t.x + o.x
	t.y = t.y + o.y
	t.z = t.z + o.z
}

func (t *Vec3) addScaled(o *Vec3, s float64) {
	t.x = t.x + o.x*s
	t.y = t.y + o.y*s
	t.z = t.z + o.z*s
}

func (t *Vec3) dot(o *Vec3) float64 {
	return t.x*o.x + t.y*o.y + t.z*o.z
}

func (t *Vec3) crossInto(o *Vec3, out *Vec3) {
	out.x = t.y*o.z - t.z*o.y
	out.y = t.z*o.x - t.x*o.z
	out.z = t.x*o.y - t.y*o.x
}

func (t *Vec3) norm() float64 {
	return math.Sqrt(t.dot(t))
}

func main() {
	const dt = 0.0000001
	p := vec3(0, 0, 0)
	v := vec3(1, 2, 3)
	g := vec3(0, -0.0000001, 0)
	kick := vec3(0, 0, 0)
	energy := 0.0
	for i := int32(0); i < N; i++ {
		v.add(g)
		v.crossInto(g, kick)
		v.addScaled(kick, 0.001)
		p.addScaled(v, dt)
		energy = energy + 0.5*v.dot(v) + p.norm()*dt
	}
	fmt.Println(p.x)
	fmt.Println(p.y)
	fmt.Println(p.z)
	fmt.Println(energy)
}

File: bench/vec3.rs
// Rust twin of vec3.ts: the same methods on a Box<Vec3> (heap objects, like
// the arena objects in Nish), same expression order.
const N: i32 = 50000000; // bench:n

struct Vec3 {
    x: f64,
    y: f64,
    z: f64,
}

impl Vec3 {
    fn new(x: f64, y: f64, z: f64) -> Box<Vec3> {
        Box::new(Vec3 { x, y, z })
    }
    fn add(&mut self, o: &Vec3) {
        self.x = self.x + o.x;
        self.y = self.y + o.y;
        self.z = self.z + o.z;
    }
    fn add_scaled(&mut self, o: &Vec3, s: f64) {
        self.x = self.x + o.x * s;
        self.y = self.y + o.y * s;
        self.z = self.z + o.z * s;
    }
    fn dot(&self, o: &Vec3) -> f64 {
        self.x * o.x + self.y * o.y + self.z * o.z
    }
    fn cross_into(&self, o: &Vec3, out: &mut Vec3) {
        out.x = self.y * o.z - self.z * o.y;
        out.y = self.z * o.x - self.x * o.z;
        out.z = self.x * o.y - self.y * o.x;
    }
    fn norm(&self) -> f64 {
        self.dot(self).sqrt()
    }
}

fn main() {
    let dt = 0.0000001;
    let mut p = Vec3::new(0.0, 0.0, 0.0);
    let mut v = Vec3::new(1.0, 2.0, 3.0);
    let g = Vec3::new(0.0, -0.0000001, 0.0);
    let mut kick = Vec3::new(0.0, 0.0, 0.0);
    let mut energy = 0.0;
    for _ in 0..N {
        v.add(&g);
        v.cross_into(&g, &mut kick);
        v.add_scaled(&kick, 0.001);
        p.add_scaled(&v, dt);
        energy = energy + 0.5 * v.dot(&v) + p.norm() * dt;
    }
    println!("{}\n{}\n{}\n{}", p.x, p.y, p.z, energy);
}
